### PR TITLE
storage support for TiCDC

### LINKED ISSUE #3963: Support setting `StorageVolumes`, `AdditionalVolumes`, `AdditionalVolumeMounts` for TiCDC
## Feature Request

**Is your feature request related to a problem? Please describe:**

Refer to the doc [here](https://docs.pingcap.com/zh/tidb/dev/troubleshoot-ticdc#%E5%90%8C%E6%AD%A5%E4%BB%BB%E5%8A%A1%E4%B8%AD%E6%96%AD%E5%B0%9D%E8%AF%95%E5%86%8D%E6%AC%A1%E5%90%AF%E5%8A%A8%E5%90%8E-ticdc-%E5%8F%91%E7%94%9F-oom%E5%A6%82%E4%BD%95%E5%A4%84%E7%90%86), we may need to support mounting volumes to the TiCDC Pod.
Refer to the logic in TiDB:
https://github.com/pingcap/tidb-operator/blob/master/pkg/manager/member/tidb_member_manager.go#L616-L619
**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### LINKED ISSUE #3996: Handle the PVC  during scaling of TiCDC
## Feature Request

**Is your feature request related to a problem? Please describe:**

We will support configuring PVCs for TiCDC in https://github.com/pingcap/tidb-operator/issues/3963, so we have to handle the PVCs during scaling in and scaling out, refer to https://github.com/pingcap/tidb-operator/blob/master/pkg/manager/member/tikv_scaler.go.
**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve?

- close #3963
- close #3996 

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Support setting StorageVolumes, AdditionalVolumes, AdditionalVolumeMounts for TiCDC
```

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@DanielZhangQD @july2993 PTAL

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 1]
```diff
@@ -1476,6 +1476,14 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 		fromTc.Spec.TiDB.Replicas = 2
 		fromTc.Spec.TiCDC.Replicas = 3
 
+		fromTc.Spec.TiCDC.StorageVolumes = []v1alpha1.StorageVolume{
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply tests/e2e/tidbcluster/tidbcluster.go]: https://github.com/pingcap/tidb-operator/pull/4004/commits/27980f153d7ae874f3bc28d7cc11bccc994820eb

File: pkg/manager/member/ticdc_scaler.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+			//
+			// The ticdc pod starts for a while, but has not synced its status, and then the pod becomes not ready.
+			// Here we wait for 5 resync periods to ensure that the status of this ticdc pod has been synced.
+			// After this period of time, if there is still no information about this ticdc in TidbCluster status,
+			// then we can be sure that this ticdc has never been added to the tidb cluster.
+			// So we can scale in this ticdc pod safely.
+			resetReplicas(newSet, oldSet)
+			return fmt.Errorf("TiCDC %s/%s is not ready, wait for 5 resync periods to sync its status", ns, podName)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/manager/member/ticdc_scaler.go]: yeah, removed in https://github.com/pingcap/tidb-operator/pull/4004/commits/27980f153d7ae874f3bc28d7cc11bccc994820eb

File: pkg/manager/member/ticdc_scaler.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	resetReplicas(newSet, oldSet)
+
+	klog.Infof("scaling in ticdc statefulset %s/%s, ordinal: %d (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinal, replicas, deleteSlots.List())
+	// We need to remove member from cluster before reducing statefulset replicas
+	var podName string
+	switch meta.(type) {
+	case *v1alpha1.TidbCluster:
+		podName = ordinalPodName(v1alpha1.TiCDCMemberType, tcName, ordinal)
+	default:
+		return fmt.Errorf("ticdcScaler.ScaleIn: failed to convert cluster %s/%s", meta.GetNamespace(), meta.GetName())
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/manager/member/ticdc_scaler.go]: https://github.com/pingcap/tidb-operator/pull/4004/commits/27980f153d7ae874f3bc28d7cc11bccc994820eb

File: pkg/manager/member/ticdc_scaler.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	return s.SyncAutoScalerAnn(meta, oldSet)
+}
+
+// ScaleOut scales out of the statefulset.
+func (s *ticdcScaler) ScaleOut(meta metav1.Object, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	_, ordinal, replicas, deleteSlots := scaleOne(oldSet, newSet)
+	resetReplicas(newSet, oldSet)
+	obj, ok := meta.(runtime.Object)
+	if !ok {
+		return fmt.Errorf("cluster[%s/%s] can't conver to runtime.Object", meta.GetNamespace(), meta.GetName())
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/manager/member/ticdc_scaler.go]: https://github.com/pingcap/tidb-operator/pull/4004/commits/27980f153d7ae874f3bc28d7cc11bccc994820eb

File: pkg/manager/member/ticdc_scaler.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+// Scale scales in or out of the statefulset.
+func (s *ticdcScaler) Scale(meta metav1.Object, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	scaling, _, _, _ := scaleOne(oldSet, newSet)
+	if scaling > 0 {
+		return s.ScaleOut(meta, oldSet, newSet)
+	} else if scaling < 0 {
+		return s.ScaleIn(meta, oldSet, newSet)
+	}
+	// we only sync auto scaler annotations when we are finishing syncing scaling
+	return s.SyncAutoScalerAnn(meta, oldSet)
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/manager/member/ticdc_scaler.go]: https://github.com/pingcap/tidb-operator/pull/4004/commits/27980f153d7ae874f3bc28d7cc11bccc994820eb

File: pkg/util/util.go [hunk 6]
```diff
@@ -180,6 +180,8 @@ func GetAutoScalingOutSlots(tc *v1alpha1.TidbCluster, memberType v1alpha1.Member
 		l = label.AnnTiKVAutoScalingOutOrdinals
 	case v1alpha1.TiDBMemberType:
 		l = label.AnnTiDBAutoScalingOutOrdinals
+	case v1alpha1.TiCDCMemberType:
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/util/util.go]: https://github.com/pingcap/tidb-operator/pull/4004/commits/27980f153d7ae874f3bc28d7cc11bccc994820eb

File: pkg/manager/member/ticdc_scaler.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+			return err
+		}
+	}
+
+	setReplicasAndDeleteSlots(newSet, replicas, deleteSlots)
+	return nil
+}
+
+// SyncAutoScalerAnn would reclaim the auto-scaling out slots if the target pod is no longer existed
+func (s *ticdcScaler) SyncAutoScalerAnn(meta metav1.Object, actual *apps.StatefulSet) error {
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/manager/member/ticdc_scaler.go]: https://github.com/pingcap/tidb-operator/pull/4004/commits/27980f153d7ae874f3bc28d7cc11bccc994820eb

File: pkg/label/label.go [hunk 8]
```diff
@@ -120,6 +120,8 @@ const (
 	AnnTiKVAutoScalingOutOrdinals = "tikv.tidb.pingcap.com/scale-out-ordinals"
 	// AnnTiDBAutoScalingOutOrdinals describe the tidb pods' ordinal list which is created by auto-scaling out
 	AnnTiDBAutoScalingOutOrdinals = "tidb.tidb.pingcap.com/scale-out-ordinals"
+	// AnnTiCDCAutoScalingOutOrdinals describe the ticdc pods' ordinal list which is created by auto-scaling out
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/label/label.go]: https://github.com/pingcap/tidb-operator/pull/4004/commits/27980f153d7ae874f3bc28d7cc11bccc994820eb

File: pkg/manager/member/ticdc_scaler.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+	resetReplicas(newSet, oldSet)
+	obj, ok := meta.(runtime.Object)
+	if !ok {
+		return fmt.Errorf("cluster[%s/%s] can't conver to runtime.Object", meta.GetNamespace(), meta.GetName())
+	}
+	klog.Infof("scaling out ticdc statefulset %s/%s, ordinal: %d (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinal, replicas, deleteSlots.List())
+	skipReason, err := s.deleteDeferDeletingPVC(obj, v1alpha1.TiCDCMemberType, ordinal)
+	if err != nil {
+		return err
+	} else if len(skipReason) != 1 || skipReason[ordinalPodName(v1alpha1.TiCDCMemberType, meta.GetName(), ordinal)] != skipReasonScalerPVCNotFound {
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/manager/member/ticdc_scaler.go:64]: How about testing and update (if necessary) in another PR later?

**Comment by csuzhangxc** [comment]
@DanielZhangQD PTAL again

File: pkg/manager/member/ticdc_scaler.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply pkg/manager/member/ticdc_scaler.go:64]: OK.

File: pkg/manager/member/ticdc_scaler.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+		klog.Errorf("cluster[%s/%s] can't convert to runtime.Object", meta.GetNamespace(), meta.GetName())
+		return nil
+	}
+	klog.Infof("scaling out ticdc statefulset %s/%s, ordinal: %d (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinal, replicas, deleteSlots.List())
+	skipReason, err := s.deleteDeferDeletingPVC(obj, v1alpha1.TiCDCMemberType, ordinal)
+	if err != nil {
+		return err
+	} else if len(skipReason) != 1 || skipReason[ordinalPodName(v1alpha1.TiCDCMemberType, meta.GetName(), ordinal)] != skipReasonScalerPVCNotFound {
+		// wait for all PVCs to be deleted
+		return controller.RequeueErrorf("ticdc.ScaleOut, cluster %s/%s ready to scale out, wait for next round", meta.GetNamespace(), meta.GetName())
```
Comment (DanielZhangQD) [inline pkg/manager/member/ticdc_scaler.go:66]: Please print out the skipReason here.

File: pkg/manager/member/ticdc_scaler.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+// Scale scales in or out of the statefulset.
+func (s *ticdcScaler) Scale(meta metav1.Object, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	scaling, _, _, _ := scaleOne(oldSet, newSet)
+	if scaling > 0 {
+		return s.ScaleOut(meta, oldSet, newSet)
+	} else if scaling < 0 {
+		return s.ScaleIn(meta, oldSet, newSet)
+	}
+	// we only sync auto scaler annotations when we are finishing syncing scaling
```
Comment (DanielZhangQD) [inline pkg/manager/member/ticdc_scaler.go:47]: Remove this line.

**Review by DanielZhangQD — COMMENTED** [review]
Rest LGTM

File: pkg/manager/member/ticdc_scaler.go (hunk 11 above)
Comment (csuzhangxc) [reply pkg/manager/member/ticdc_scaler.go]: https://github.com/pingcap/tidb-operator/pull/4004/commits/472f2e93cc0c6be310c36d8ba78c049cdf023190

File: pkg/manager/member/ticdc_scaler.go (hunk 10 above)
Comment (csuzhangxc) [reply pkg/manager/member/ticdc_scaler.go]: https://github.com/pingcap/tidb-operator/pull/4004/commits/472f2e93cc0c6be310c36d8ba78c049cdf023190

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by july2993 — APPROVED** [review]

**Pull request merged**